osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs

```diff
@@ -360,62 +329,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)
             }
         }
 
-        private void startCountdown(MultiplayerCountdown countdown, Func<Task> continuation)
```

Comment:
Removing this means that countdowns are no longer testable in "flow" tests (`TestSceneMultiplayer`).

Reply:
Is the goal here to test it ad-hoc interactively? Would you be against making an "InteractiveMultiplayer" test scene which uses the online (dev) server for said purposes?

Reply:
Let's not use online anything, please. I am still traumatised by the CI test failures when online stuff was used (simple things like timeouts to httpbin or a.ppy.sh), and is the reason why this has been reimplemented client-side.

The test is testing actual functionality.

Reply:
For sure, I was proposing *only* for use in interactive visual tests (ie during development). Not replacing usage in this test.

Reply:
That said, I'm fine with leaving said code in place until I have a proposal which fully replaces it, if that's preferred. I removed it mostly to gauge response (given that it's not used anywhere else and one of the most complex pieces of `TestMultiplayerClient`.

Reply:
So your thinking is as long as it isn't used in automated testing, it isn't required? I guess I could get along with that.

Personally I don't find this test scene all that useful interactively, opting to instead test the full client + server flow via in-game.

Reply:
My current line of thought is that:

- We need to test at component boundaries more often. Every expectation of the component should be tested in isolation (sometimes we do that, other times we don't, this is mostly a proposal for going forward).
- For "flow" tests, where more than one components are involved in cross-component interactions, they should be tested together in an isolated scene, but still not require the multiplayer client/server.
- We need test coverage of `MultiplayerClient` on its own, with the input and output expectations of it. This will hopefully reduce the perceived need for some of the tests in `TestSceneMultiplayer`.

I'm planning to continue taking this a step at a time (working through every scene that is using `OnlinePlayTestScene`) to reduce the cross-dependencies of tests, as that seems to be the biggest point of stress in the system.

As you say, for development it's easy to spin up a new tests scene and get the online support when required to test the "overall picture", but I think we need to do that less in tests and try and isolate at boundaries more, splitting up responsibilities of individual pieces to reduce the overall complexity (and make refactoring the pieces and related tests easier).

Still formulating things and doing some reading on good testing practices, but that's where I am as of today.